cupy/core/reduction.pxi

```diff
@@ -175,24 +175,94 @@ cpdef list _get_inout_args(
 
 @util.memoize(for_each_device=True)
 def _get_simple_reduction_function(
-        routine, params, args_info, in_arg_dtype, out_arg_dtype, out_types,
+        map_expr, reduce_expr, post_map_expr, reduce_type,
+        params, args_info, types,
         name, block_size, identity, input_expr, output_expr, _preamble,
         options):
-    reduce_type = routine[3]
-    if reduce_type is None:
-        reduce_type = _get_typename(out_types[0])
+    params = _get_kernel_params(params, args_info)
 
-    t = (_get_typename(in_arg_dtype), _get_typename(out_arg_dtype))
-    type_preamble = 'typedef %s type_in0_raw; typedef %s type_out0_raw;' % t
+    type_preamble = '\n'.join(
+        'typedef %s %s;' % (_get_typename(v), k)
+        for k, v in types)
 
-    params = _get_kernel_params(params, args_info)
     return _get_simple_reduction_kernel(
         name, block_size, reduce_type, params, identity,
-        routine[0], routine[1], routine[2],
+        map_expr, reduce_expr, post_map_expr,
         type_preamble, input_expr, output_expr, _preamble, options)
 
 
-class simple_reduction_function(object):
+class _AbstractReductionKernel:
```

Comment:
Is it possible to make the base class a `cdef` one?

Reply:
Thank you for suggestion.
Done in 5339ade1718679e8864f9803b298abbed1eee280.